From ae88531fcf6e7d899e31b5029959d916012ba735 Mon Sep 17 00:00:00 2001
From: Benjamin Reikowski <reikowski@dresearch-fe.de>
Date: Wed, 9 Dec 2015 12:52:35 +0100
Subject: [PATCH] fixed hikirk PHY driver issue HYP-11185

saving 1000BaseT register an extra time to keep advertisement configuration

Signed-off-by: Benjamin Reikowski <reikowski@dresearch-fe.de>
---
 drivers/net/phy/micrel.c     |  4 +++-
 drivers/net/phy/phy_device.c | 16 ++++++++++++++++
 2 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/drivers/net/phy/micrel.c b/drivers/net/phy/micrel.c
index 3ad8ca7..323ba48 100644
--- a/drivers/net/phy/micrel.c
+++ b/drivers/net/phy/micrel.c
@@ -72,6 +72,8 @@
 
 #define PS_TO_REG				200
 
+int genphy_resume_fixed(struct phy_device *phydev); /* HYP-11185 */
+
 struct kszphy_type {
 	u32 led_mode_reg;
 	u16 interrupt_level_mask;
@@ -724,7 +726,7 @@ static struct phy_driver ksphy_driver[] = {
 	.ack_interrupt	= kszphy_ack_interrupt,
 	.config_intr	= kszphy_config_intr,
 	.suspend	= genphy_suspend,
-	.resume		= genphy_resume,
+	.resume		= genphy_resume_fixed, /* HYP-11185 */
 	.driver		= { .owner = THIS_MODULE, },
 }, {
 	.phy_id		= PHY_ID_KSZ8873MLL,
diff --git a/drivers/net/phy/phy_device.c b/drivers/net/phy/phy_device.c
index 3fc91e8..837cfd5 100644
--- a/drivers/net/phy/phy_device.c
+++ b/drivers/net/phy/phy_device.c
@@ -1177,6 +1177,22 @@ int genphy_resume(struct phy_device *phydev)
 }
 EXPORT_SYMBOL(genphy_resume);
 
+int genphy_resume_fixed(struct phy_device *phydev) /* HYP-11185 */
+{
+        int value,rescue;
+
+        mutex_lock(&phydev->lock);
+
+        rescue = phy_read(phydev, 0x09); /* save 1000BaseT register content due to a race between suspend and reset */
+        value = phy_read(phydev, MII_BMCR);
+        phy_write(phydev, MII_BMCR, value & ~BMCR_PDOWN);
+        phy_write(phydev, 0x09, rescue); /* write back saved content */
+
+        mutex_unlock(&phydev->lock);
+        return 0;
+}
+EXPORT_SYMBOL(genphy_resume_fixed);
+
 static int gen10g_resume(struct phy_device *phydev)
 {
 	return 0;
-- 
2.1.4

